### PR TITLE
Changed `HttpLoggingPolicy` to ensure a response is logged even if its body is never consumed by a client.

### PR DESCRIPTION
**Pull request by vcolin7**
Currently, the `HttpLoggingPolicy.LoggingHttpResponse` class we use for logging a response will only log when either its `Flux<ByteBuffer> getBody()` or `BinaryData getBodyAsBinaryData()` methods get called (I assume to avoid eagerly buffering the response body). This causes a situation where policies such as `BearerTokenAuthenticationPolicy` don't log the response because they don't care about accessing the body but [only the headers](https://github.com/Azure/azure-sdk-for-java/blob/af51239e7ac5254cd1f6272f247f33eea8340f53/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java#L155).
 
We're changing the logic in `HttpLoggingPolicy` to log the response no matter if the body is accessed or not. If the `HttpLogLevel` does not include the body (e.g. `HEADERS`), then it will not be consumed eagerly, otherwise we will include it in the log. We assume body logging is used in non-production scenarios, as it is more resource intensive, and it usually is enabled only when debugging issues.

As for tests, there are still a few failing ones, so I'll spend some time this week fixing that scenario.

### CODE REVIEW

**Review by lmolkova — COMMENTED** [review]
We should ensure we close/consume bodies whenever possible because, orthogonal to logging, this signals underlying HTTP client implementation that response is consumed and it can return the underlying connection to the pool where it can be used for other requests.

Not closing the response body leads (orthogonally to logging) to connection staying 'busy' waiting for someone to consume the response, creating a bottleneck for the application.

While I have no objections eagerly buffering when request/response body logging is enabled, I believe we need to solve a bigger problem (if) we leak connections.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java [hunk 1]
```diff
@@ -315,20 +324,35 @@ public Mono<HttpResponse> logResponse(ClientLogger logger, HttpResponseLoggingCo
...
+
                 if (shouldBodyBeLogged(contentTypeHeader, contentLength)) {
                     // shouldBodyBeLogged ensures contentLength is not null and within limits.
                     int contentLengthInt = contentLength.intValue();
-                    return Mono.just(new LoggingHttpResponse(response, logBuilder, logger, contentLengthInt,
-                        contentTypeHeader, prettyPrintBody));
+
+                    AccessibleByteArrayOutputStream stream = new AccessibleByteArrayOutputStream(contentLengthInt);
+
+                    response.getBody().subscribe(byteBuffer -> {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:341]: I'd recommend against using `subscribe` within a reactive flow anywhere that isn't a sink-based call chain. This doesn't block and may result in race conditions or unexpected behaviors.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:1]: Should update both request and response body to be eager. While we can make some assumptions that the request body will be sent at some point it'll be easier to handle both in the same way, even if it means making request body logging less performant.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java [hunk 2]
```diff
@@ -315,20 +324,35 @@ public Mono<HttpResponse> logResponse(ClientLogger logger, HttpResponseLoggingCo
...
+                        try {
+                            ImplUtils.writeByteBufferToStream(byteBuffer.duplicate(), stream);
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:343]: We can simplify our usage to `BinaryData` only by using `BinaryData.writeTo(AsynchronousByteChannel)` and creating an `AccessibleByteArrayOutputStream`-based `AsynchronousByteChannel`.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java [hunk 3]
```diff
@@ -364,20 +381,22 @@ public HttpResponse logResponseSync(ClientLogger logger, HttpResponseLoggingCont
...
 
             if (httpLogDetailLevel.shouldLogBody()) {
                 String contentTypeHeader = response.getHeaderValue(HttpHeaderName.CONTENT_TYPE);
+
                 if (shouldBodyBeLogged(contentTypeHeader, contentLength)) {
-                    // shouldBodyBeLogged ensures contentLength is not null and within limits.
-                    int contentLengthInt = contentLength.intValue();
-                    return new LoggingHttpResponse(response, logBuilder, logger, contentLengthInt, contentTypeHeader,
-                        prettyPrintBody);
+                    logBuilder.addKeyValue(LoggingKeys.BODY_KEY, prettyPrintIfNeeded(logger, prettyPrintBody,
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:393]: Need to buffer the HTTP response here as well.

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**